fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java

```diff
@@ -528,6 +530,58 @@ public ScalarOperator visitTimestampArithmeticExpr(TimestampArithmeticExpr node,
+            LogicalPlan subqueryPlan = SubqueryUtils.getLogicalPlan(session, cteContext, columnRefFactory,
+                    relation, builder.getExpressionMapping());
+
+            if (relation instanceof SelectRelation &&
+                    !subqueryPlan.getCorrelation().isEmpty() && ((SelectRelation) relation).hasAggregation()) {
+                throw new SemanticException(
+                        "Unsupported correlated in predicate subquery with grouping or aggregation");
+            }
+            return subqueryPlan;
+        }
+
+        @Override
+        public ScalarOperator visitMultiInPredicate(MultiInPredicate node, Context context) throws SemanticException {
+            if (!(node.getChild(node.getTupleSize()) instanceof Subquery)) {
+                throw new SemanticException("multi-column IN predicate is only supported for subquery expressions");
+
+            }
+            QueryStatement queryStatement = ((Subquery) node.getChild(node.getTupleSize())).getQueryStatement();
+            LogicalPlan subqueryPlan = getSubqueryPlan(queryStatement);
+            List<ColumnRefOperator> leftCorrelationColumns = Lists.newArrayList();
+            List<ScalarOperator> leftExprs = Lists.newArrayList();
+            for (int i = 0; i < node.getTupleSize(); ++i) {
+                ScalarOperator leftColRef = SqlToScalarOperatorTranslator.translate(node.getChild(i),
+                        builder.getExpressionMapping(), leftCorrelationColumns, columnRefFactory);
+                if (leftCorrelationColumns.size() > 0) {
+                    throw new SemanticException("Unsupported complex nested in-subquery");
+                }
+                leftExprs.add(leftColRef);
+            }
+            List<ColumnRefOperator> rightColRefs = subqueryPlan.getOutputColumn();
+            if (rightColRefs.size() != leftExprs.size()) {
+                throw new SemanticException("subquery must return the same number of columns as provided by the IN predicate");
+            }
+            ScalarOperator inPredicateOperator = new MultiInPredicateOperator(node.isNotIn(), leftExprs, rightColRefs);
+            ColumnRefOperator outputPredicateRef = columnRefFactory.create(inPredicateOperator,
+                    inPredicateOperator.getType(), inPredicateOperator.isNullable());
+            ((Subquery) node.getChild(node.getTupleSize())).setUseSemiAnti(context.useSemiAnti);
+
+            LogicalApplyOperator applyOperator = LogicalApplyOperator.builder().setOutput(outputPredicateRef)
+                    .setSubqueryOperator(inPredicateOperator)
+                    .setCorrelationColumnRefs(subqueryPlan.getCorrelation())
+                    .setUseSemiAnti(context.useSemiAnti).build();
+
+            SubqueryOperator subqueryOperator = new SubqueryOperator(rightColRefs.get(0).getType(), queryStatement,
+                    applyOperator, subqueryPlan.getRootBuilder());
+
```

Comment:
here is to deceive the optimizer, the type of subquery can't be the first column’s type in actually, and it's need don't handle MultiInOperator cast in ImplicitCastRule, I think it's a trick and will cause other problems... you need more test and comment it

Reply:
should i add a special TupleType?
I agree it's a bit of a trick, but this subquery operator should be fully removed by one of the "earlier" logical rules (like QuantifiedApply2JoinRule). We could add a check that the MultiInPredicateOperator is not in the query plan after those subquery rewriting rules.
I modified the ImplicitCastRule to throw an exception if the MultiInPredicate is encountered.

Reply:
I think don't need add speical TupleType now, but you'd better add comment here. btw, it's a good idea to throw an exception in ImplicitCastRule

Reply:
comment regarding the type is added